src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -5376,7 +5376,7 @@ public ulong GetChannelKeyword(EventChannel channel, ulong channelKeyword = 0)
             if (!channelTab.TryGetValue((int)channel, out ChannelInfo? info))
             {
                 // If we were not given an explicit channel, allocate one.
-                if (channelKeyword != 0)
+                if (channelKeyword == 0)
```

Comment:
Do we need to put this new keyword into `channelTab`?

Reply:
If EventAttribute.Channel is not EventChannel.None, I believe EventDescriptor.Keywords will need to include ChannelInfo.Keywords. Now if the event source class defines a nested type `Channels` and lists the channel there, I think AddChannel called by AddProviderEnumKind will take care of that. But if the event source class instead uses a predefined channel such as EventChannel.Operational and does not redeclare it in `Channels`, then I'm not sure whether that will work without further changes.

If the event source class defines multiple event methods with the same EventAttribute.Channel but different channel keywords in EventAttribute.Keywords, I believe that is an error.

If I understand correctly, Windows Event Log computes the channel keyword from the order of `channel` elements in the manifest, and ChannelInfo.Keywords needs to match that. I'm not sure whether that is really documented anywhere. Perhaps Microsoft decided that the requirement does not need to be documented if the Message Compiler satisfies it when generating code from an instrumentation manifest.

Reply:
I'm not sure anyone on the runtime team knows about the specifics of channels. We may have to reach out to the Windows team for guidance

Reply:
I responded before I fully understood the question 😄 . The code has some more complicated dependencies.

`GetChannelKeyword` does not add it to `channelTab`, but it's only called in two places - `AddChannel` (which does immediately add it to `channelTab`) and `CreateManifestAndDescriptors`.

`CreateManifestAndDescriptors` should only ever be called once, so it should be ok. The risk is if that invariant is broken or someone adds another call to`GetChannelKeyword` that doesn't make sure it's cached in channelTab first, you could end up in a state where the same channel has two difference keywords associated with it. So it's bad design and potentially will have an issue down the road, but I think it's correct as it is now.